### PR TITLE
Gh570

### PR DESCRIPTION
**Pull request by josephwright**
*Pull requests in this repository are intended for LaTeX Team members only.*

# Internal housekeeping

## Status of pull request

- Feedback wanted 
- Under development
- Ready to merge

## Checklist of required changes before merge will be approved
- [x] Test file(s) added
- [x] Version and date string updated in changed source files
- [x] Relevant `\changes` entries in source included
- [x] Relevant `changes.txt` updated
- [x] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

### CODE REVIEW

File: base/doc/ltnews34.tex [hunk 1]
```diff
@@ -376,7 +376,15 @@ \subsection{???}
...
 \section{Code improvements}
 
-
+\subsection{Detect ``\cs{endfoo}'' when using \cs{NewDocumentEnvironment}}
+
+The \cs{newenvironment} command has always checked that both \cs{foo} and
+\cs{endfoo} do not exist before creating a \texttt{foo} environment. In
+contrast, \cs{NewDocumentEnvironment} has only to date checked for \cs{foo};
+this reflects the fact that historically the code was designed around the
+target of an entirely new format. The behaviour of \cs{NewDocumentEnvironment}
```
Comment (FrankMittelbach) [inline base/doc/ltnews34.tex:385]: US spelling?

File: base/ltcmd.dtx [hunk 2]
```diff
@@ -3599,6 +3599,13 @@
     with~an~environment~that~already~has~a~definition. \\ \\
     The~existing~definition~of~'#1'~will~not~be~altered.
   }
+\msg_new:nnnn { cmd } { env-end-already-defined }
+  { End~of~environment~'#1'~already~defined. }
```
Comment (FrankMittelbach) [inline base/ltcmd.dtx:3603]: Similar Errors end in "!"

**Review by FrankMittelbach — APPROVED** [review]
Give or take the conflicts that need resolving  and some minor comments this looks fine

**Comment by FrankMittelbach** [comment]
Actually I wonder ... the way this now works if you try `\NewDocumentEnvironment{quote}...`you get 2 error messages not one, right? Might be nicer if there is only one like with `\newenvironment`.

**Comment by josephwright** [comment]
> Actually I wonder ... the way this now works if you try `\NewDocumentEnvironment{quote}...`you get 2 error messages not one, right? Might be nicer if there is only one like with `\newenvironment`.

Huh? There should be only one message as `\quote` exists, so we never look at `\endquote`.

**Comment by FrankMittelbach** [comment]
> > Actually I wonder ... the way this now works if you try `\NewDocumentEnvironment{quote}...`you get 2 error messages not one, right? Might be nicer if there is only one like with `\newenvironment`.
> 
> Huh? There should be only one message as `\quote` exists, so we never look at `\endquote`.

ah good, then I misunderstood an earlier remark (I hadn't checked the code, sorry). That is fine then and in fact even better than what is done in `\newenvironment`

**Pull request merged**